### ISSUE TITLE
Implement a Tasker extension for invoking synchronization

### ISSUE THREAD

**Issue by novoid** [orgzly/orgzly-android#231]
Hi,

The well known Android tool [Tasker](http://tasker.dinglisch.net/) offers the possibility to invoke external hooks of Android apps that provide such a "plugin" mechanism. (I am sure there is a specific term for this I don't know yet.)

It would solve many people's problems when Orglzy would come with such an Tasker plugin extension so that my Tasker is able to do a sync whenever it suits my requirements.

Background info:

In my case, the beta auto-sync function does not work for me because I do have huge Org-mode files. When I activate "sync whenever Orgzly is started", I have to wait a couple of minutes(!) before I could actually use Orglzy because during the sync, Orglzy behaves not normal. I can't search, I can't navigate.

Therefore, I'd prefer a sync in the morning, when I leave the house (loosing my WiFi network). I am sure that other people do have other preferences such as entering the company's WiFi, or every other hour or or or. Instead of implementing complex "when to sync" functionality into Orgzly, users could benefit from the awesome possibilities of Tasker.

**Comment by nevenz** [comment]
Plugin would be cool, in the meantime, you can do this by sending an intent - see [this comment](https://github.com/orgzly/orgzly-android/issues/8#issuecomment-288946032).

I've tried it now by creating a new action `System -> Send Intent`:

**Action**
`com.orgzly.intent.action.SYNC_START`

**Package**
`com.orgzly`

**Class**
`com.orgzly.android.sync.SyncService`

**Target**
`Service`

**Comment by novoid** [comment]
While the workaround using the intent worked for a couple of months, I now got:

`code 3: java.lang.IllegalStateException: Not allowed to start service Intent { act=com.orgzly.intent.action.SYNC_START cmp=com.orgzly/.android.sync.SyncService }: app is in background uid null.` which results in not synchronizing.

Current setup:
- Tasker v5.1m
- Orgzly 1.6.6
- Android 8.1.0

I found a helpful explanation for the changed behavior with https://github.com/orgzly/orgzly-android/issues/315#issuecomment-381704021 which requires to switch from "startservice" to "start-foreground-service" starting with Android Oreo. Issue #315 further mentions solutions/workarounds for the app "Automate" but not for Tasker. 

However, https://github.com/claiire updated https://github.com/orgzly/orgzly-android/issues/8#issuecomment-304738355 where the process for Tasker was explained in the first place:

> - Open up tasker and go to the Profiles tab.
> - Click the + at the bottom-center of the screen and select Time
> - Leave From alone. Uncheck To, check Repeat and put in the update interval you'd like.
> - Now that you've set up your Context for updating, go back to the Profile screen and there should be an Enter task box hovering over your new profile. Hit New Task, give it a name like "Sync Orgzly" or something, and hit the checkmark. This should open the Task Edit screen.
> - Add a new Action to the Task with the plus button at the bottom of the screen. Select Code > Run Shell on the popup.
> - In the Command field, write `am startservice --user 0 -a com.orgzly.intent.action.SYNC_START com.orgzly/com.orgzly.android.sync.SyncService`
> Tasker should now sync orgzly every x minutes/hours/whatever you used. Probably. I'm bad at this stuff and new to it.

This works for me (again).

**Comment by novoid** [comment]
Hi @nevenz,

What do I have to enter as command? Because `am startservice --user 0 -a com.orgzly.intent.action.SYNC_START com.orgzly/com.orgzly.android.ActionReceiver` did result in an error 255 with no sync process initiated.

**Comment by nevenz** [comment]
> What do I have to enter as command?

```
adb shell am broadcast -a com.orgzly.intent.action.SYNC_START com.orgzly/.android.ActionReceiver
```

I should probably document this somewhere.

**Comment by novoid** [comment]
Still an error. Are you sure about the `com.orgzly/.android.ActionReceiver` part?

**Comment by nevenz** [comment]
> Still an error. Are you sure about the com.orgzly/.android.ActionReceiver part?

Yeah, first two work for me. What's the error, the same?

**Comment by novoid** [comment]
Grml. Yesterday, all I got was errors (code 255 AFAIR, nothing too specific).

Today, when I re-started the task for copying the error messages, without any change, it works. The shutdown/reboot over night must have changed something IMO.

Sorry for the fuzz.

**Comment by Nelsonlittle** [comment]
What seems to work for me is:

Tasker -> Send Intent

Action: `com.orgzly.intent.action.SYNC_START`
Package: `com.orgzly`
Class: `com.orgzly.android.ActionReceiver`

This is probably  what @nevenz mentioned before

**Comment by novoid** [comment]
Thanks @Nelsonlittle for clarification - I did not get that before.

I just checked again:

- Tasker -> Send Intent
- Action: `com.orgzly.intent.action.SYNC_START`
- Package: `com.orgzly`
- Class: `com.orgzly.android.ActionReceiver`

... does not work at my side.

- Tasker > Run Shell
- Command: `am startservice --user 0 -a com.orgzly.intent.action.SYNC_START com.orgzly/com.orgzly.android.sync.SyncServic`

... currently works on my side.

**Comment by nevenz** [comment]
> I just checked again:
> 
> Tasker -> Send Intent
> Action: com.orgzly.intent.action.SYNC_START
> Package: com.orgzly
> Class: com.orgzly.android.ActionReceiver
> 
> ... does not work at my side.

What Target did you use? It needs to be Broadcast Receiver.

**Comment by novoid** [comment]
Oh, thanks for clarification. I used the default one which was not Broadcast Receiver.

However, changing it to Broadcast Receiver did not result in Orgzly syncing.

**Comment by nevenz** [comment]
> However, changing it to Broadcast Receiver did not result in Orgzly syncing.

Hmm, I'm not sure what could be the issue. It's hard to debug without seeing logs. If you have `adb` you might be able to see what's happening. Although there might not be enough useful infromation without debugging enabled in the app. I do see at least:

```
06-17 21:36:26.153  1648  1665 I ActivityManager: Start proc 22497:com.orgzly/u0a104 for broadcast com.orgzly/.android.ActionReceiver
```

when I run Tasker's task.

**Comment by rieje** [comment]
I could not find intent or plugin in documentation--if intents work, shouldn't they be documented? Only just started using Tasker and Orgzly was my immediate thought. Surprised this isn't a higher priority.

**Comment by nevenz** [comment]
I agree sync intents should be documented. If someone wants to take a shot, /help page is at https://github.com/orgzly/documentation/blob/master/miscellaneous.org.

> Is SYNC_STOP ever necessary?

It's not, it's there for completeness.

**Comment by codygman** [comment]
I'm trying to send a sync event to orgzly when a folder is modified but having trouble. I started trying to send a sync event and "sync..." popup with a delay for just a file and that doesn't seem to work either.

I do have a this UX for when I switch to orgzly though and it limits the amount of conflicts I get drastically. The only difference from the orgzly "sync on resume/leave app" is that there is a popup that stays on screen for 1 second.

**Comment by nevenz** [comment]
@codygman: I'm not sure I understand the exact issue. Is it about initiating the sync, or sync conflict after that?

**Comment by ugurbolat** [comment]
These are the parameters to send syncservice command:
Package: com.orgzly
Service Class: com.orgzly.sync.SyncService
Action: com.orgzly.intent.action.SYNC_START
"Start ub foreground state" enabled (without this,I got "app is in background uid null" error)

**Comment by novoid** [comment]
> @codygman: I'm not sure I understand the exact issue. Is it about initiating the sync, or sync conflict after that?

It's about initiating the sync.

**Comment by sh-zam** [comment]
Since the recent update the command has changed. Intent should be sent to `BroadcastReciever class: com.orgzly.android.ActionReceiver`.

**Comment by nulinspiratie** [comment]
Since the update I haven't been able to get Tasker to sync properly. I've been using the parameters of @ugurbolat and the updated intent of @sh-zam .

Has anyone recently managed to get Tasker to send a sync command? If so could you write down the parameters used?

**Comment by falematte** [comment]
Do we have any working setup avaible?
